core/src/al/miniaudio_device.cpp

```diff
@@ -0,0 +1,238 @@
+
+using namespace cubos::core::al;
+
+MiniaudioBuffer::MiniaudioBuffer(const void* data, size_t dataSize)
+{
+    if (ma_decoder_init_memory(data, dataSize, nullptr, &mDecoder) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to initialize Decoder from data");
+        abort();
+    }
+}
+
+size_t MiniaudioBuffer::getLength()
+{
+    ma_uint64 lengthInPCMFrames;
+    ma_result result = ma_decoder_get_length_in_pcm_frames(&mDecoder, &lengthInPCMFrames);
+
+    if (result != MA_SUCCESS)
+    {
+        CUBOS_ERROR("Failed to get the length of audio in PCM frames.");
+        return 0;
+    }
+
+    // Calculate the length in seconds: Length in PCM frames divided by the sample rate.
+    return static_cast<size_t>(lengthInPCMFrames) / mDecoder.outputSampleRate;
+}
+
+MiniaudioBuffer::~MiniaudioBuffer()
+{
+    ma_decoder_uninit(&mDecoder);
+};
+
+MiniaudioSource::MiniaudioSource()
+{
+    if (ma_engine_init(nullptr, &mEngine) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to initialize miniaudio engine.");
+        abort();
+    }
+}
+
+MiniaudioSource::~MiniaudioSource()
+{
+    ma_sound_uninit(&mSound);
+    ma_engine_uninit(&mEngine);
+}
+
+void MiniaudioSource::setBuffer(Buffer buffer)
+{
+    if (ma_sound_init_from_data_source(&mEngine, &buffer->mDecoder, 0, nullptr, &mSound) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to initialize sound from buffer.");
+        abort();
+    }
+}
+
+void MiniaudioSource::setPosition(const glm::vec3& position)
+{
+    ma_sound_set_position(&mSound, position.x, position.y, position.z);
+}
+
+void MiniaudioSource::setVelocity(const glm::vec3& velocity)
+{
+    ma_sound_set_velocity(&mSound, velocity.x, velocity.y, velocity.z);
+}
+
+void MiniaudioSource::setGain(float gain)
+{
+    ma_sound_set_volume(&mSound, gain);
+}
+
+void MiniaudioSource::setPitch(float pitch)
+{
+    ma_sound_set_pitch(&mSound, pitch);
+}
+
+void MiniaudioSource::setLooping(bool looping)
+{
+    ma_sound_set_looping(&mSound, static_cast<ma_bool32>(looping));
+}
+
+void MiniaudioSource::setRelative(bool relative)
+{
+    relative ? ma_sound_set_positioning(&mSound, ma_positioning_relative)
+             : ma_sound_set_positioning(&mSound, ma_positioning_absolute);
+}
+
+void MiniaudioSource::setMaxDistance(float maxDistance)
+{
+    ma_sound_set_max_distance(&mSound, maxDistance);
+}
+
+void MiniaudioSource::setMinDistance(float minDistance)
+{
+    ma_sound_set_min_distance(&mSound, minDistance);
+}
+
+void MiniaudioSource::setCone(float innerAngle, float outerAngle, float outerGain = 1.0F)
+{
+    ma_sound_set_cone(&mSound, innerAngle, outerAngle, outerGain);
+}
+
+void MiniaudioSource::setConeDirection(const glm::vec3& direction)
+{
+    ma_sound_set_direction(&mSound, direction.x, direction.y, direction.z);
+}
+
+void MiniaudioSource::play()
+{
+    if (ma_sound_start(&mSound) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to start sound.");
+        abort();
```

Comment:
```suggestion
        CUBOS_ERROR("Failed to start sound");
```
This doesn't seem fatal!